src/internal/storage/fileset/shard.go

```diff
@@ -7,49 +7,53 @@ import (
 	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index"
 )
 
+func (s *Storage) EstimateShards(ctx context.Context, ids []ID, cb index.ShardCallback) error {
+	fs, err := s.Open(ctx, ids)
+	if err != nil {
+		return err
+	}
+	err = fs.Shard(ctx, cb)
+	return errors.EnsureStack(err)
+}
+
 // Shard shards the file set into path ranges.
 // TODO This should be extended to be more configurable (different criteria
 // for creating shards).
-func (s *Storage) Shard(ctx context.Context, ids []ID, cb ShardCallback) error {
-	fs, err := s.Open(ctx, ids)
+func (s *Storage) Shard(ctx context.Context, ids []ID, pathRange *index.PathRange, cb index.ShardCallback) error {
+	fs, err := s.Open(ctx, ids, index.WithRange(pathRange))
 	if err != nil {
 		return err
 	}
-	return shard(ctx, fs, s.shardSizeThreshold, s.shardCountThreshold, cb)
+	return shard(ctx, fs, s.shardSizeThreshold, s.shardCountThreshold, pathRange, cb)
 }
 
-// ShardCallback is a callback that returns a path range for each shard.
-type ShardCallback func(*index.PathRange) error
-
 // shard creates shards (path ranges) from the file set streams being merged.
 // A shard is created when the size of the files is greater than or equal to the passed
 // in size threshold or the file count is greater than or equal to the passed in count threshold.
 // For each shard, the callback is called with the path range for the shard.
-func shard(ctx context.Context, fs FileSet, sizeThreshold, countThreshold int64, cb ShardCallback) error {
+func shard(ctx context.Context, fs FileSet, sizeThreshold, countThreshold int64, basePathRange *index.PathRange, cb index.ShardCallback) error {
+	pathRange := &index.PathRange{
+		Lower: basePathRange.Lower,
+	}
 	var size, count int64
-	pathRange := &index.PathRange{}
 	if err := fs.Iterate(ctx, func(f File) error {
-		// A shard is created when we have encountered more than shardThreshold content bytes.
 		if size >= sizeThreshold || count >= countThreshold {
+			pathRange.Upper = f.Index().Path
```

Comment:
I am considering preventing this from happening in general. I could just track the previous while sharding and bypass the normal sharding criteria when a duplicate path appears. This could result with larger shards in certain cases, but I would prefer that over having to ensure correctness when this occurs. I think this could only be pathological if a file was written to by a lot of datums, which would be incorrect usage of Pachyderm.

Reply:
Either way is fine with me